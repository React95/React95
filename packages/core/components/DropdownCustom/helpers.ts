import {
  DropdownCustomOption,
  DropdownCustomOptionObject,
} from './DropdownCustom';

const isOptionObject = (
  option: DropdownCustomOption,
): option is DropdownCustomOptionObject =>
  typeof option === 'object' && option !== null && 'value' in option;

const convertOptionToObjects = (
  option: DropdownCustomOption,
): DropdownCustomOptionObject => {
  if (isOptionObject(option)) {
    return option;
  }
  return { value: option };
};

export const convertOptionsToObjects = (options: DropdownCustomOption[]) =>
  options.map(convertOptionToObjects);
