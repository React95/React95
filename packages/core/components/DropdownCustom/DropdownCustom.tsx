import React, {
  forwardRef,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { useClickAway } from 'react-use';
import {
  CaretButton,
  Option,
  Options,
  Overlay,
  Select,
  SelectedValue,
  Wrapper,
} from './DropdownCustom.styles';

import { convertOptionsToObjects } from './helpers';

type Value = string | number;
export type DropdownCustomOptionObject = {
  value: Value;
  onClick?: (value: Value, index: number) => void;
};
export type DropdownCustomOption = Value | DropdownCustomOptionObject;

type PropsWithHTMLAttributes<T, U> = Omit<U, keyof T> & T;

export type DropdownCustomProps = PropsWithHTMLAttributes<
  {
    options: DropdownCustomOption[];
    defaultOpen?: boolean;
    defaultIndex?: number;
    maxHeight?: CSSStyleDeclaration['maxHeight'];
    onChange?: (value: Value, index: number) => void;
  },
  React.HTMLAttributes<HTMLDivElement>
>;

const DropdownCustom = forwardRef<HTMLDivElement, DropdownCustomProps>(
  (
    {
      options,
      defaultOpen = false,
      defaultIndex,
      maxHeight,
      onChange,
      ...rest
    },
    ref,
  ) => {
    const [isOpen, setOpen] = useState(defaultOpen);
    const [selectedIndex, setSelectedIndex] = useState(defaultIndex);
    const [hoveredIndex, setHoveredIndex] = useState<number | null>(null);
    const [optionObjects, setOptionObjects] = useState<
      DropdownCustomOptionObject[]
    >(convertOptionsToObjects(options));

    useEffect(() => {
      setOptionObjects(convertOptionsToObjects(options));
    }, [options]);

    const optionsRef = useRef<HTMLDivElement>(null);
    useClickAway(optionsRef, () => {
      setOpen(false);
      setHoveredIndex(null);
    });

    const resetHoveredIndex = useCallback(() => {
      setHoveredIndex(null);
    }, []);

    const handleSelectClick = useCallback(() => {
      setOpen(!isOpen);
      if (isOpen) {
        resetHoveredIndex();
      }
    }, [isOpen]);

    const handleOptionClick = useCallback((index: number) => {
      setSelectedIndex(index);
      resetHoveredIndex();
      const option = optionObjects[index];

      onChange?.(option.value, index);
      option.onClick?.(option.value, index);
    }, []);

    const handleOptionHover = useCallback((index: number) => {
      setHoveredIndex(index);
    }, []);

    return (
      <Wrapper>
        <Select {...rest} ref={ref} onClick={handleSelectClick}>
          <Overlay>
            <SelectedValue>
              {selectedIndex !== undefined &&
                optionObjects[selectedIndex].value}
            </SelectedValue>
            <CaretButton />
          </Overlay>
          <Options
            isOpen={isOpen}
            onMouseLeave={resetHoveredIndex}
            maxHeight={maxHeight}
            ref={optionsRef}
          >
            {optionObjects.map((option, index) => (
              <Option
                key={option.value}
                isHighlighted={
                  index === hoveredIndex ||
                  (index === selectedIndex && hoveredIndex === null)
                }
                onClick={() => handleOptionClick(index)}
                onMouseEnter={() => handleOptionHover(index)}
              >
                {option.value}
              </Option>
            ))}
          </Options>
        </Select>
      </Wrapper>
    );
  },
);

export default DropdownCustom;
