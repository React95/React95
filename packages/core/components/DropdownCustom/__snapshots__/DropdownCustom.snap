// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownCustom /> Snapshots should match snapshot with options prop 1`] = `
.c0 {
  position: relative;
  width: 200px;
  height: 20px;
}

.c1 {
  position: relative;
  outline: none;
  border: none;
  border-radius: 0;
  width: 100%;
  height: 20px;
  padding: 3px;
  color: #000000;
  background-color: #ffffff;
  border-left: 1px solid;
  border-left-color: #868a8e;
  border-top: 1px solid;
  border-top-color: #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb, inset 1px 1px 0 0 #000000, 0.5px 0.5px 0 0.5px #ffffff;
  -webkit-appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c5 {
  display: none;
  position: absolute;
  top: calc(100% + 1px);
  left: -1px;
  width: calc(100% + 2px);
  border: 1px solid #000000;
}

.c6 {
  background: #fff;
  color: #000000;
  padding: 1px 2px;
  font-size: 12px;
  border: 1px solid transparent;
}

.c6:hover {
  background: #000e7a;
  color: #fff;
  border: 1px dotted #d2d2d2;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  width: 100%;
  height: 100%;
  pointer-events: none;
}

.c3 {
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  font-size: 12px;
}

.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 17px;
  height: 17px;
  font-size: 14px;
  line-height: 1.1;
  -webkit-transform: translate(-3px,-2px);
  -ms-transform: translate(-3px,-2px);
  transform: translate(-3px,-2px);
  background-color: #c3c7cb;
  box-shadow: inset 0.5px 0.7px 0px 0.7px #c3c7cb, inset -1px 0px 0 1px #868a8e, inset 1.5px 1.5px 0px 1.5px #ffffff;
  border-right: 1px solid;
  border-bottom: 1px solid;
  background-image: url('test-file-stub');
  background-position: 60% 60%;
  background-repeat: no-repeat;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        />
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <div
          class="c6"
        />
        <div
          class="c6"
        >
          C:\\Windows
        </div>
        <div
          class="c6"
        >
          C:\\Windows/System32
        </div>
      </div>
    </div>
  </div>
</div>
`;
